sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java

```diff
@@ -2002,6 +2006,11 @@ static String getExtractDestinationUri(String extractDestinationDir) {
       return toBuilder().setFormatFunction(formatFunction).build();
     }
 
+    /** Formats the user's type into a {@link TableRow} to be written to an error collector. */
+     public Write<T> withFailsafeFormatFunction(SerializableFunction<T, TableRow> formatFunction) {
```

Comment:
Does this have to write TableRows? Why not also have the option to return the user type itself? (perhaps it's a lot of trouble to implement given the existing types?)
What are some examples where users may want to recover a different tablerow than the one that was sent to BQ?

Reply:
Currently the design is for the ErrorContainer to return TableRow's (or BigQueryError<TableRow>).  My worry is that it wouldn't be backwards compatible to make this generic, though I agree it would be more valuable.

I think this question is important and is my only big open question around this design

Reply:
I've been looking into this feature and it will require a number of additional changes.  IMO it makes more sense to merge this first feature, then add a second issue / PR for the dynamic typing as it appears to be a relatively complex addon